add dynamic config metrics

mtaufen opened the pull request:
This PR exports config-releated metrics from the Kubelet.
The Guages for active, assigned, and last-known-good config can be used
to identify config versions and produce aggregate counts across several
nodes. The error-reporting Gauge can be used to determine whether a node
is experiencing a config-related error, and to prodouce an aggregate
count of nodes in an error state.

https://github.com/kubernetes/features/issues/281

```release-note
The Kubelet now exports metrics that report the assigned (node_config_assigned), last-known-good (node_config_last_known_good), and active (node_config_active) config sources, and a metric indicating whether the node is experiencing a config-related error (node_config_error). The config source metrics always report the value 1, and carry the node_config_name, node_config_uid, node_config_resource_version, and node_config_kubelet_key labels, which identify the config version. The error metric reports 1 if there is an error, 0 otherwise.
```

crassirostris commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -283,3 +352,18 @@ func (pc *podAndContainerCollector) Collect(ch chan<- prometheus.Metric) {
+
+func SetAssignedNodeConfigSource(val string) {
+	// TODO(mtaufen): will this eliminate previous metric, or generate a parallel one?
+	AssignedNodeConfigSource.WithLabelValues(val).Set(1)
```
This approach, implemented like this, is going to result in memory leak, because old values are not GC'd, as you noted in the comment

crassirostris replied:
@loburm @x13n FYI, this is the usecase for string metrics

crassirostris commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -40,13 +43,31 @@ const (
+	AssignedNodeConfigSourceKey = "assigned_node_config_source"
+	// AssignedNodeConfigSourceOkayKey's metric indicates whether the currently assigned config source and the config it contains is valid
+	AssignedNodeConfigSourceOkayKey       = "assigned_node_config_source_okay"
+	AssignedNodeConfigSourceStatusTrue    = 1
```
Please use string values, performance considerations in this case do not outweigh the decreased readability

mtaufen replied:
As in, just report a string value in another label key-value pair?

Since prometheus doesn't support string valued metrics:
https://prometheus.io/docs/concepts/metric_types/
https://github.com/prometheus/prometheus/issues/2227

crassirostris commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -40,13 +43,31 @@ const (
+	// AssignedNodeConfigSourceKey's metric reports the config source the Kubelet believes is currently assigned to it
+	AssignedNodeConfigSourceKey = "assigned_node_config_source"
+	// AssignedNodeConfigSourceOkayKey's metric indicates whether the currently assigned config source and the config it contains is valid
+	AssignedNodeConfigSourceOkayKey       = "assigned_node_config_source_okay"
```
Nit: IMHO `ok` is more widely used than `okay`

mtaufen replied:
fair enough

crassirostris commented on pkg/kubelet/kubeletconfig/controller.go:
```diff
@@ -108,9 +109,20 @@ func NewController(defaultConfig *kubeletconfig.KubeletConfiguration,
 	utillog.Infof("starting controller")
 
+	// set metrics when all is said and done
+	var using string
```
`using` is a bad name, is says nothing and spans a huge scope

Also, if you change the logic below later, it may be used uninitialized

crassirostris replied:
I don't understand that global defer at all, why not leave setting metrics local?

crassirostris commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -101,7 +122,7 @@ var (
 		},
 	)
-	// Metrics of remote runtime operations.
+	// Metrics for remote runtime operations
```
Mostly I see a slow convergence towards comments being complete sentences, with full stop at the end and so on.

crassirostris commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -40,13 +43,31 @@ const (
 	DevicePluginAllocationLatencyKey = "device_plugin_alloc_latency_microseconds"
+	// Metric keys for node config sources
+	// AssignedNodeConfigSourceKey's metric reports the config source the Kubelet believes is currently assigned to it
+	AssignedNodeConfigSourceKey = "assigned_node_config_source"
```
Why the `assigned_node` prefix? It's little big confusing, why not just `config_source` or `dynamic_config_source`?

mtaufen replied:
There are three config references you'd want the Kubelet to report:
1. What it believes its currently assigned source is.
2. What it believes its last-known-good source is.
3. Which of the above two it's actually using. 

`NodeConfigSource` is the type of the `ConfigSource` field on the `Node` object and is used to point the Kubelet at a config source, hence `node_config_source`.

Could do:
1. `node_config_source_assigned`
2. `node_config_source_last_known_good`
3. `node_config_source_using` (or some name other than `using`)

crassirostris commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -40,13 +43,31 @@ const (
+	AssignedNodeConfigSourceStatusUnknown = 0
+	AssignedNodeConfigSourceStatusFalse   = -1
+	// LastKnownGoodNodeConfigSourceKey's metric indicates what the Kubelet is treating as it's last-known-good config source
+	LastKnownGoodNodeConfigSourceKey = "last_known_good_node_config_source"
```
Try to make these hierarchical, that helps, e.g. `dynamic_config_last_known_source`

crassirostris reviewed (commented):
Overall approach SGTM

mtaufen commented:
Thanks @crassirostris for taking a look at this

mtaufen commented:
/retest

mtaufen committed:
add dynamic config metrics

This PR exports config-releated metrics from the Kubelet.
The Guages for active, assigned, and last-known-good config can be used
to identify config versions and produce aggregate counts across several
nodes. The error-reporting Gauge can be used to determine whether a node
is experiencing a config-related error, and to prodouce an aggregate
count of nodes in an error state.

mtaufen commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -47,6 +51,17 @@ const (
+	ActiveConfigKey               = "node_config_active"
+	LastKnownGoodConfigKey        = "node_config_last_known_good"
+	ConfigErrorKey                = "node_config_error"
+	ConfigNameLabelKey            = "node_config_name" // this is a fully resolved name, e.g. an API path or a URL, "Name" is short but a little confusing name for this key...
```
@dashpole any naming suggestions for this?

dashpole replied:
SelfLink is the URL representing an object in ObjectMeta.  Maybe "node_config_self_link"?  This should be equivalent to the self link for the configmap, right?

mtaufen replied:
@dashpole and I settled on node_config_source, as it accurately identifies what the corresponding value describes while being reasonably generic ("local", or a SelfLink, or maybe some URL in the future).

mtaufen commented:
/retest

mtaufen commented:
/status approved-for-milestone

mtaufen commented:
/remove-lifecycle stale

mtaufen commented:
/retest

dashpole commented:
/lgtm

dchen1107 commented:
/lgtm
